yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,412 @@
+
+        else:
+            diff_x = ((node.left_edge[0] + node.right_edge[0]) / 2 - posx)
+            diff_x *= diff_x
+            diff_y = ((node.left_edge[1] + node.right_edge[1]) / 2 - posy)
+            diff_y *= diff_y
+            diff_z = ((node.left_edge[2] + node.right_edge[2]) / 2 - posz)
+            diff_z *= diff_z
+
+            q_ij = math.sqrt((diff_x + diff_y + diff_z) / (hsml*hsml))
+
+            buff[node.leaf_id] += prefactor * self.kernel(q_ij)
+
+    def interpolate_sph_octs(self, np.float64_t[:] buff, np.float64_t[:] posx,
+                             np.float64_t[:] posy, np.float64_t[:] posz,
+                             np.float64_t[:] pmass, np.float64_t[:] pdens,
+                             np.float64_t[:] hsml, np.float64_t[:] field,
+                             kernel_name="cubic"):
+
+        self.kernel = get_kernel_func(kernel_name)
+
+        cdef int i
+        cdef double prefactor
+
+        for i in range(posx.shape[0]):
+            prefactor = pmass[i] / pdens[i] / hsml[i]**3
+            prefactor *= field[i]
+
+            self.smooth_onto_leaves(buff, posx[i], posy[i], posz[i], hsml[i],
+                                    prefactor, self.c_tree.root)
+
+# this is a utility function which is able to catch errors when we attempt to
+# reserve too much memory
+cdef int reserve(vector[Node] * vec, int amount) except +MemoryError:
+    vec.reserve(amount)
+    return 0
+
+# NOTE: maybe move these into the PyOctree
+# TODO: move these into the pyoctree class
+# this makes the children and stores them in the tree.node
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef inline int generate_children(Octree &tree, Node * node) nogil:
+    cdef int i, j, z, k
+    cdef double dx, dy, dz
+    cdef Node temp
+
+    node.children = tree.nodes.size()
+
+    temp.parent = node.node_id
+    temp.leaf = 1
+    temp.depth = node.depth + 1
+    temp.leaf_id = 0
+
+    tree.max_depth = max(temp.depth, tree.max_depth)
+
+    dx = (node.right_edge[0] - node.left_edge[0]) / 2
+    dy = (node.right_edge[1] - node.left_edge[1]) / 2
+    dz = (node.right_edge[2] - node.left_edge[2]) / 2
+
+    z = node.children
+    for i in range(2):
+        for j in range(2):
+            for k in range(2):
+                temp.left_edge[0] = node.left_edge[0] + i*dx
+                temp.left_edge[1] = node.left_edge[1] + j*dy
+                temp.left_edge[2] = node.left_edge[2] + k*dz
+                temp.right_edge[0] = node.left_edge[0] + (i+1)*dx
+                temp.right_edge[1] = node.left_edge[1] + (j+1)*dy
+                temp.right_edge[2] = node.left_edge[2] + (k+1)*dz
+                temp.node_id = z
+                tree.nodes.push_back(temp)
+                z+=1
+    return 0
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int process_node(Octree &tree, Node * node, double * input_pos) nogil:
```

Comment:
Might this cdef function or the ones elsewhere in this file ever raise an exception? If so I think you just need to add `except -1` to the signature to make the exceptions propagate up the call stack from here. In general every cdef function should explicitly handle exceptions IMO.

Reply:
fixed